### DISCUSSION TITLE
ExpressBruteFlexible with Sentinel Redis store behaves unexpectedly

### DISCUSSION THREAD

**Discussion by mcandea04** [animir/node-rate-limiter-flexible#223] [Q&A] [answered]
Hi, having some trouble when migrating my limiters using `ExpressBruteFlexible` from Mongo to Redis store.
I noticed that now they allow only `freeRetries-1` instead of `freeRetries`, and after free retries are exhausted don't allow any more requests, even after waiting for the `minWait` (and even `maxWait`) duration, until the `lifetime` period passes.
Also I noticed only the entry with prefix `free:` is created in Redis, and no `block:` and `counter:`, like I saw it happening in Mongo (probably this is the root cause of the behaviour).

I am using `rate-limiter-flexible` v2.4.2, and have same problem with 3.0.0 too.
For the Redis connection I am using `ioredis` v4.28.5.
I have tried both with and without `{enable_offline_queue: false}` when building the Redis client, and still no luck.

This is how I build the limiter:
```
new ExpressBruteFlexible(ExpressBruteFlexible.LIMITER_TYPES.REDIS, {
		storeClient: getSimpleClient({ ...config.redis }),
		freeRetries: 5,
		minWait: 60000,
		maxWait: 300000,
})
```
and then call `.getMiddleware()` on it to obtain the middleware to use in the endpoint.

Any clue would be appreciated 🙏 Thanks in advance!

**Comment by animir** [comment]
@mcandea04 Hi, do you see some error or usual `Too many requests in this time frame.` ?

**Reply by mcandea04** [reply]
Thanks for the prompt reply @animir, just the standard message `Too many requests in this time frame.`

**Reply by animir** [reply]
@mcandea04 What `nextValidRequestDate` value is there? Is it incorrect?

**Reply by mcandea04** [reply]
@animir that's right, the `nextValidRequestDate` is also incorrect, it's about the same time as the request itself.

**Reply by animir** [reply]
@mcandea04 Could you share `getSimpleClient` and `config.redis`?

**Reply by mcandea04** [reply]
@animir, I'm not certain if I can disclose too much 😊, but `getSimpleClient` essentially creates a `RedisClient` and attaches a few listeners, like for disconnects. The `config.redis` contains `sentinels` and `name`. 
No issues with the Redis connection have been observed elsewhere in the app. As the original post mentioned, an entry with the `free:` prefix is being created. 
Is this information sufficient for further investigation?

**Reply by animir** [reply]
@mcandea04 `ExpressBruteFlexible` with Redis works in other projects, so more you provide, easier it would be to help.
Could you test it without Redis Sentinel, but with Redis?

**Reply by mcandea04** [reply]
@animir I have tested connecting directly to Redis master and not to the sentinel, and indeed the behaviour is as expected in that case😳.
Does that help, or do you still need me to share `getSimpleClient` and `config.redis`?
Unfortunately bypassing the sentinel is not an option for our setup.

**Reply by animir** [reply]
@mcandea04 Unfortunately, I've never used Redis Sentinel, and not sure what it may be. I guess It isn't supported by `rate-limiter-flexible`. What we could do here is to try to dig dipper, why Sentinel creates one key, but not another. Maybe, it stores keys on different nodes? If so, is there a way to set some parameter to make it work as usual Redis Server?

**Reply by mcandea04** [reply, marked as answer]
Hello again @animir! So apparently of all the changes that I've tried, the one that actually made the difference was NOT bypassing the Redis sentinel, but upgrading `rate-limiter-flexible` from 2.4.2 to 3.0.0. My bad there, I must have got confused somehow.
Which makes sense, since I'm using `ioredis` and the [release notes for 3.0.0](https://github.com/animir/node-rate-limiter-flexible/releases/tag/v3.0.0) do mention improved support for it.
So I think this is solved. Many thanks again for your swift support! 🙌

**Reply by animir** [reply]
@mcandea04 You're welcome.
